internal/loader/configuration_client_manager.go

```diff
@@ -0,0 +1,424 @@
+}
+
+func (manager *ConfigurationClientManager) RefreshClients() {
+	currentTime := metav1.Now()
+	if manager.ReplicaDiscoveryEnabled &&
+		currentTime.After(manager.lastFallbackClientsAttempt.Time.Add(MinimalClientRefreshInterval)) {
+		manager.lastFallbackClientsAttempt = currentTime
+		_ = manager.DiscoverFallbackClients(manager.Endpoint)
+	}
+}
+
+func (manager *ConfigurationClientManager) DiscoverFallbackClients(host string) error {
+	srvTargetHosts, err := SrvTargetHostQuery(host)
+	if err != nil {
+		return err
+	}
+
+	// Shuffle the list of SRV target hosts
+	for i := range srvTargetHosts {
+		j := rand.Intn(i + 1)
+		srvTargetHosts[i], srvTargetHosts[j] = srvTargetHosts[j], srvTargetHosts[i]
+	}
+
+	newDynamicClients := make([]*ConfigurationClientWrapper, 0)
+	for _, host := range srvTargetHosts {
+		if isValidEndpoint(host, manager.ValidDomain) {
+			targetEndpoint := "https://" + host
+			client, err := manager.newConfigurationClient(targetEndpoint)
+			if err != nil {
+				return fmt.Errorf("build fallback client look up failed")
+			}
+			newDynamicClients = append(newDynamicClients, &ConfigurationClientWrapper{
+				Endpoint:       targetEndpoint,
+				Client:         client,
+				BackOffEndTime: metav1.Time{},
+				FailedAttempts: 0,
+			})
+		}
+	}
+
+	manager.DynamicClients = newDynamicClients
+	manager.lastFallbackRefresh = metav1.Now()
+	return nil
+}
+
+func (manager *ConfigurationClientManager) UpdateClientBackoffStatus(endpoint string, successful bool) {
+	if manager.ReplicaDiscoveryEnabled != true {
+		return
+	}
+	if successful {
+		manager.ClientsFallbackStatus[endpoint].BackOffEndTime = metav1.Time{}
+		manager.ClientsFallbackStatus[endpoint].FailedAttempts = 0
+	} else {
+		manager.ClientsFallbackStatus[endpoint].FailedAttempts++
+		manager.ClientsFallbackStatus[endpoint].BackOffEndTime = metav1.Time{Time: metav1.Now().Add(calculateBackoffDuration(manager.ClientsFallbackStatus[endpoint].FailedAttempts))}
+	}
+}
+
+func (manager *ConfigurationClientManager) newConfigurationClient(endpoint string) (*azappconfig.Client, error) {
+	if manager.ConnectionString == "" {
+		return azappconfig.NewClient(endpoint, manager.Credential, clientOptionWithModuleInfo)
+	}
+
+	connectionStr := buildConnectionString(endpoint, manager.secret, manager.id)
+	if connectionStr == "" {
+		return nil, fmt.Errorf("failed to build connection string for fallback client")
+	}
+	return azappconfig.NewClientFromConnectionString(connectionStr, clientOptionWithModuleInfo)
+}
+
+func isValidEndpoint(host string, validDomain string) bool {
+	if validDomain == "" {
+		return false
+	}
+
+	return strings.HasSuffix(strings.ToLower(host), strings.ToLower(validDomain))
+}
+
+func getValidDomain(endpoint string) string {
+	url, _ := url.Parse(endpoint)
+	TrustedDomainLabels := []string{AzConfigDomainLabel, AppConfigDomainLabel}
+	for _, label := range TrustedDomainLabels {
+		index := strings.LastIndex(strings.ToLower(url.Host), strings.ToLower(label))
```

Comment:
could you some examples of `url` and expected index? I'm wondering if there is any way for customer to bypass the validation above.

Reply:
Sure, added some examples in endpoint validation test.